File: app/src/main/java/com/absinthe/anywhere_/view/OverlayView.java
package com.absinthe.anywhere_.view;

import android.annotation.SuppressLint;
import android.content.Context;
import android.content.Intent;
import android.os.Handler;
import android.os.Message;
import android.view.HapticFeedbackConstants;
import android.view.MotionEvent;
import android.view.View;
import android.view.ViewConfiguration;
import android.view.WindowManager;
import android.widget.ImageButton;
import android.widget.LinearLayout;

import androidx.annotation.NonNull;

import com.absinthe.anywhere_.services.OverlayService;
import com.absinthe.anywhere_.utils.CommandUtils;
import com.absinthe.anywhere_.utils.Logger;
import com.absinthe.anywhere_.utils.UiUtils;

public class OverlayView extends LinearLayout {

    private static final int MSG_REMOVE_WINDOW = 1001;

    private final Context mContext;
    private final WindowManager mWindowManager;
    private final int mTouchSlop;

    private WindowManager.LayoutParams mLayoutParams;
    private ImageButton ibIcon;

    private String mCommand;
    private String mPkgName;

    private boolean isClick;
    private long mStartTime = 0;
    private long mEndTime = 0;

    @SuppressLint("HandlerLeak")
    private Handler mHandler = new Handler() {
        public void handleMessage(@NonNull Message msg) {
            if (msg.what == MSG_REMOVE_WINDOW) {
                ibIcon.performHapticFeedback(HapticFeedbackConstants.LONG_PRESS);
                mContext.startService(
                        new Intent(mContext, OverlayService.class)
                                .putExtra(OverlayService.COMMAND, OverlayService.COMMAND_CLOSE)
                );
            }
        }
    };

    public OverlayView(Context context) {
        super(context);
        mContext = context;
        mWindowManager = (WindowManager) context.getSystemService(Context.WINDOW_SERVICE);
        mTouchSlop = ViewConfiguration.get(mContext).getScaledTouchSlop();
        initView();
    }

    @SuppressLint("ClickableViewAccessibility")
    private void initView() {
        int width = UiUtils.dipToPixels(mContext, 65);
        int height = UiUtils.dipToPixels(mContext, 65);
        setLayoutParams(new LinearLayout.LayoutParams(width, height));

        ibIcon = new ImageButton(mContext);
        LinearLayout.LayoutParams layoutParams = new LayoutParams(width, height);
        ibIcon.setLayoutParams(layoutParams);
        ibIcon.setBackground(null);
        addView(ibIcon);

        ibIcon.setOnClickListener(v -> {
            Logger.d("Overlay window clicked!");

            CommandUtils.execCmd(mCommand);
        });

        ibIcon.setOnTouchListener(new OnTouchListener() {

            private float lastX; //上一次位置的X.Y坐标
            private float lastY;
            private float nowX;  //当前移动位置的X.Y坐标
            private float nowY;
            private float tranX; //悬浮窗移动位置的相对值
            private float tranY;

            @Override
            public boolean onTouch(View view, MotionEvent motionEvent) {
                mLayoutParams = (WindowManager.LayoutParams) OverlayView.this.getLayoutParams();
                switch (motionEvent.getAction()) {
                    case MotionEvent.ACTION_DOWN:
                        // 获取按下时的X，Y坐标
                        lastX = motionEvent.getRawX();
                        lastY = motionEvent.getRawY();
                        Logger.d("MotionEvent.ACTION_DOWN last:", lastX, lastY);

                        isClick = false;
                        mStartTime = System.currentTimeMillis();
                        mHandler.sendEmptyMessageDelayed(MSG_REMOVE_WINDOW, 1000);
                        break;
                    case MotionEvent.ACTION_MOVE:
                        isClick = true;

                        // 获取移动时的X，Y坐标
                        nowX = motionEvent.getRawX();
                        nowY = motionEvent.getRawY();
                        Logger.d("MotionEvent.ACTION_MOVE now:", nowX, nowY);

                        // 计算XY坐标偏移量
                        tranX = nowX - lastX;
                        tranY = nowY - lastY;
                        Logger.d("MotionEvent.ACTION_MOVE tran:", tranX, tranY);

                        if (tranX * tranX + tranY * tranY > mTouchSlop * mTouchSlop) {
                            mHandler.removeMessages(MSG_REMOVE_WINDOW);
                        }

                        // 移动悬浮窗
                        mLayoutParams.x -= tranX;
                        mLayoutParams.y += tranY;
                        //更新悬浮窗位置
                        mWindowManager.updateViewLayout(OverlayView.this, mLayoutParams);
                        //记录当前坐标作为下一次计算的上一次移动的位置坐标
                        lastX = nowX;
                        lastY = nowY;

                        break;
                    case MotionEvent.ACTION_UP:
                        mEndTime = System.currentTimeMillis();
                        Logger.d("Touch period =", (mEndTime - mStartTime));

                        isClick = (mEndTime - mStartTime) > 0.2 * 1000L;
                        mHandler.removeMessages(MSG_REMOVE_WINDOW);
                        break;
                }
                return isClick;
            }
        });
    }

    public String getCommand() {
        return mCommand;
    }

    public void setCommand(String mCommand) {
        this.mCommand = mCommand;
    }

    public String getPkgName() {
        return mPkgName;
    }

    public void setPkgName(String mPkgName) {
        this.mPkgName = mPkgName;
        ibIcon.setImageDrawable(UiUtils.getAppIconByPackageName(mContext, mPkgName));
    }
}

File: app/src/main/java/com/absinthe/anywhere_/view/home/OverlayView.kt
package com.absinthe.anywhere_.view.home

import android.annotation.SuppressLint
import android.content.Context
import android.view.*
import android.widget.LinearLayout
import com.absinthe.anywhere_.AwContextWrapper
import com.absinthe.anywhere_.model.database.AnywhereEntity
import com.absinthe.anywhere_.services.overlay.IOverlayService
import com.absinthe.anywhere_.utils.UxUtils
import com.absinthe.anywhere_.utils.handler.Opener
import com.absinthe.anywhere_.viewbuilder.entity.OverlayBuilder
import com.absinthe.libraries.utils.extensions.dp
import timber.log.Timber

@SuppressLint("ViewConstructor")
class OverlayView(context: Context, private val binder: IOverlayService, windowLayoutParams: WindowManager.LayoutParams) : LinearLayout(context) {

    var entity: AnywhereEntity = AnywhereEntity()
        set(value) {
            field = value
            mBuilder.ivIcon.setImageDrawable(UxUtils.getAppIcon(context, value, 65.dp))
            mBuilder.tvName.text = value.appName
        }

    private val mWindowManager: WindowManager = context.getSystemService(Context.WINDOW_SERVICE) as WindowManager
    private val mTouchSlop: Int = ViewConfiguration.get(context).scaledTouchSlop
    private var mLayoutParams = windowLayoutParams

    private lateinit var mBuilder: OverlayBuilder
    private var isClick = false
    private var mStartTime: Long = 0
    private var mEndTime: Long = 0

    private val removeWindowTask = Runnable {
        mBuilder.ivIcon.performHapticFeedback(HapticFeedbackConstants.LONG_PRESS)
        binder.closeOverlay(entity)
    }

    init {
        initView()
    }

    @SuppressLint("ClickableViewAccessibility")
    private fun initView() {
        mBuilder = OverlayBuilder(context, this)

        mBuilder.ivIcon.setOnClickListener {
            Opener.with(AwContextWrapper(context)).load(entity).open()
        }
        mBuilder.ivIcon.setOnTouchListener(object : OnTouchListener {
            //Last x, y position = 0f
            private var lastX = 0f
            private var lastY = 0f

            //Current x, y position = 0f
            private var nowX = 0f
            private var nowY = 0f

            //悬浮窗移动位置的相对值 = 0f
            private var tranX = 0f
            private var tranY = 0f

            override fun onTouch(view: View, motionEvent: MotionEvent): Boolean {

                when (motionEvent.action) {
                    MotionEvent.ACTION_DOWN -> {
                        // 获取按下时的X，Y坐标
                        lastX = motionEvent.rawX
                        lastY = motionEvent.rawY
                        Timber.d("MotionEvent.ACTION_DOWN last: %f %f", lastX, lastY)
                        isClick = false
                        mStartTime = System.currentTimeMillis()
                        postDelayed(removeWindowTask, 1000)
                    }
                    MotionEvent.ACTION_MOVE -> {
                        isClick = true

                        // 获取移动时的X，Y坐标
                        nowX = motionEvent.rawX
                        nowY = motionEvent.rawY
                        Timber.d("MotionEvent.ACTION_MOVE now: %f %f", nowX, nowY)

                        // 计算XY坐标偏移量
                        tranX = nowX - lastX
                        tranY = nowY - lastY
                        Timber.d("MotionEvent.ACTION_MOVE tran: %f %f", tranX, tranY)
                        if (tranX * tranX + tranY * tranY > mTouchSlop * mTouchSlop) {
                            removeCallbacks(removeWindowTask)
                        }

                        // 移动悬浮窗
                        mLayoutParams.apply {
                            x -= tranX.toInt()
                            y += tranY.toInt()
                        }
                        //更新悬浮窗位置
                        mWindowManager.updateViewLayout(this@OverlayView, mLayoutParams)
                        //记录当前坐标作为下一次计算的上一次移动的位置坐标
                        lastX = nowX
                        lastY = nowY
                    }
                    MotionEvent.ACTION_UP -> {
                        mEndTime = System.currentTimeMillis()
                        Timber.d("Touch period = %d", mEndTime - mStartTime)
                        isClick = mEndTime - mStartTime > 0.2 * 1000L
                        removeCallbacks(removeWindowTask)
                    }
                }
                return isClick
            }
        })
    }
}